File: frontend/app/lib/utils/practice_draft_storage.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class PracticeDraft {
  final String lessonSlug;
  final int currentQuestionIndex;
  final Map<String, dynamic> selectedAnswers;
  final Map<String, dynamic> writtenAnswers;
  final DateTime updatedAt;
  final int totalQuestions;
  final int version;

  PracticeDraft({
    required this.lessonSlug,
    required this.currentQuestionIndex,
    required this.selectedAnswers,
    required this.writtenAnswers,
    required this.updatedAt,
    required this.totalQuestions,
    this.version = 1,
  });

  Map<String, dynamic> toJson() => {
        'lessonSlug': lessonSlug,
        'currentQuestionIndex': currentQuestionIndex,
        'selectedAnswers': selectedAnswers,
        'writtenAnswers': writtenAnswers,
        'updatedAt': updatedAt.toIso8601String(),
        'totalQuestions': totalQuestions,
        'version': version,
      };

  static PracticeDraft? fromJson(Map<String, dynamic> json) {
    if (json['version'] != 1) return null;
    final lessonSlug = json['lessonSlug'];
    final currentQuestionIndex = json['currentQuestionIndex'];
    final totalQuestions = json['totalQuestions'];
    final updatedAt = json['updatedAt'];

    if (lessonSlug is! String || lessonSlug.trim().isEmpty) return null;
    if (currentQuestionIndex is! int || currentQuestionIndex < 0) return null;
    if (totalQuestions is! int || totalQuestions < 0) return null;
    if (updatedAt is! String) return null;

    return PracticeDraft(
      lessonSlug: lessonSlug,
      currentQuestionIndex: currentQuestionIndex,
      selectedAnswers: _stringMap(json['selectedAnswers']),
      writtenAnswers: _stringMap(json['writtenAnswers']),
      updatedAt: DateTime.tryParse(updatedAt) ?? DateTime.fromMillisecondsSinceEpoch(0),
      totalQuestions: totalQuestions,
    );
  }

  static Map<String, dynamic> _stringMap(dynamic value) {
    if (value is Map) {
      return Map<String, dynamic>.fromEntries(
        value.entries.map(
          (entry) => MapEntry(entry.key.toString(), entry.value),
        ),
      );
    }
    return <String, dynamic>{};
  }
}

class PracticeDraftStorage {
  static const int version = 1;

  static String buildKey({
    required String classLevel,
    required String subjectSlug,
    required String lessonSlug,
  }) {
    final scope = [_normalize(classLevel), _normalize(subjectSlug)]
        .where((part) => part.isNotEmpty)
        .join('_');
    return 'practice_draft_mobile_${scope}_$lessonSlug';
  }

  static Future<PracticeDraft?> load(String key) async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(key);
    if (raw == null || raw.isEmpty) return null;

    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map<String, dynamic>) return null;
      return PracticeDraft.fromJson(decoded);
    } catch (_) {
      return null;
    }
  }

  static Future<void> save(String key, PracticeDraft draft) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(key, jsonEncode(draft.toJson()));
    } catch (_) {
      // Ignore storage failures.
    }
  }

  static Future<void> clear(String key) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(key);
    } catch (_) {
      // Ignore storage failures.
    }
  }

  static String _normalize(String value) {
    final trimmed = value.trim();
    if (trimmed.isEmpty) return '';
    if (trimmed == '+1' || trimmed.toLowerCase() == 'plus1') return 'Class_11';
    if (trimmed == '+2' || trimmed.toLowerCase() == 'plus2') return 'Class_12';
    final compact = trimmed.replaceAll(RegExp(r'\s+'), '_');
    return compact[0].toUpperCase() + compact.substring(1);
  }
}

File: frontend/web/src/utils/practiceDraftStorage.js
const DRAFT_PREFIX = 'practice_draft_web_'
const DRAFT_VERSION = 1

function hasWindow() {
  return typeof window !== 'undefined' && typeof localStorage !== 'undefined'
}

function normalizeSegment(value) {
  const trimmed = String(value ?? '').trim()
  if (!trimmed) return ''
  if (trimmed === '+1' || trimmed.toLowerCase() === 'plus1') return 'Class_11'
  if (trimmed === '+2' || trimmed.toLowerCase() === 'plus2') return 'Class_12'
  const compact = trimmed.replace(/\s+/g, '_')
  return compact.charAt(0).toUpperCase() + compact.slice(1)
}

export function buildPracticeDraftKey({ classLevel, subjectSlug, lessonSlug }) {
  const scope = [classLevel, subjectSlug]
    .map(normalizeSegment)
    .filter(Boolean)
    .join('_')
  return `${DRAFT_PREFIX}${scope}_${normalizeSegment(lessonSlug)}`
}

export function readPracticeDraft(key) {
  if (!hasWindow() || !key) return null

  try {
    const raw = localStorage.getItem(key)
    if (!raw) return null
    const parsed = JSON.parse(raw)
    if (!parsed || typeof parsed !== 'object') return null
    if (parsed.version !== DRAFT_VERSION) return null
    if (typeof parsed.lessonSlug !== 'string' || !parsed.lessonSlug.trim()) return null
    if (!Number.isInteger(parsed.totalQuestions) || parsed.totalQuestions < 0) return null
    if (!Number.isInteger(parsed.currentQuestionIndex) || parsed.currentQuestionIndex < 0) return null
    return parsed
  } catch {
    return null
  }
}

export function writePracticeDraft(key, draft) {
  if (!hasWindow() || !key) return
  try {
    localStorage.setItem(key, JSON.stringify(draft))
  } catch {
    // Ignore quota and serialization failures.
  }
}

export function clearPracticeDraft(key) {
  if (!hasWindow() || !key) return
  try {
    localStorage.removeItem(key)
  } catch {
    // Ignore storage failures.
  }
}

export function clampIndex(index, total) {
  if (!Number.isInteger(index)) return 0
  if (!Number.isInteger(total) || total <= 0) return 0
  return Math.max(0, Math.min(total - 1, index))
}
